src/Cli/dotnet/Commands/Run/FileBasedAppSourceEditor.cs

```diff
@@ -0,0 +1,223 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.DotNet.Cli.Commands.Run;
+
+/// <summary>
+/// A helper to perform edits of file-based app C# source files (e.g., updating the directives).
+/// </summary>
+internal sealed class FileBasedAppSourceEditor
+{
+    public SourceFile SourceFile
+    {
+        get;
+        private set
+        {
+            field = value;
+
+            // Make sure directives are reloaded next time they are accessed.
+            Directives = default;
+        }
+    }
+
+    public ImmutableArray<CSharpDirective> Directives
+    {
+        get
+        {
+            if (field.IsDefault)
+            {
+                field = LoadDirectives(SourceFile);
+                Debug.Assert(!field.IsDefault);
+            }
+
+            return field;
+        }
+        private set
+        {
+            field = value;
+        }
+    }
+
+    public required string NewLine { get; init; }
+
+    private FileBasedAppSourceEditor() { }
+
+    public static FileBasedAppSourceEditor Load(SourceFile sourceFile)
+    {
+        return new FileBasedAppSourceEditor
+        {
+            SourceFile = sourceFile,
+            Directives = LoadDirectives(sourceFile),
```

Comment:
nit: if Directives is just initialized based on value of `SourceFile`, would it be cleaner to delete the setter for it and just rely on it being initialized on access?